### ISSUE TITLE
add possibility to jump between modes

### ISSUE THREAD

**Issue by bbernhard** [bbernhard/imagemonkey-core#131]
idea originated in #127 

just some brainstorming to see if we can come up with a good workflow that allows us to switch between different modes. I would see that as an intermediate solution until we have a "pro mode" in place that unifies all the modes in one view. 

Option #1: 

A dedicated button which switches to the selected mode (by opening it in a new browser tab?)

![mode_switch](https://user-images.githubusercontent.com/5710631/41180703-585d259c-6b6f-11e8-950e-8b1fef6190e9.gif)

Option #2: 

A custom right mousebutton menu that opens when one clicks on an image. That probably doesn't work on mobile/desktop appliances (but maybe we could use a long touchscreen press there?). 

I kind of like the idea of having a custom menu, as it keeps the main screen "clean". I also tried to play with different colors and moved the button to different positions, but no matter where i put it, it either looked misplaced or too eye-catching. :/ But not sure if a custom right mousebutton menu is easy to discover? 

@dobkeratops what do you think?

**Comment by dobkeratops** [comment]
would it be possible to use the existing top edge buttons i.e. when switching between between'add labels','annotate','validate label','validate annotations' , keep the current image if the current screen has one (i.e. if using any one of those) , and of course advance it on 'done',

besides that, the 'goto..' idea seems good;

the other thing that would be awesome is being able to get to any of those screens after selecting an image from 'explore'

**Comment by bbernhard** [comment]
> would it be possible to use the existing top edge 'tabs'/buttons i.e. when switching between between'add labels','annotate','validate label','validate annotations' normally , just keep the current image if the current screen has one (i.e. if using any one of those) , and of course advance it on 'done',

I thought quite a while about it, but I think it might not be always obvious to the user when a new random image gets picked and when the image is kept. 

> besides that, the 'goto..' idea (option 1) seems good... maybe keep the right click available for actual label editing (+/- scribbles, draw/erase or draw/select...)

good point!

> the other thing that would be awesome is being able to get to any of those screens after selecting an image from 'explore'. 

good idea, thanks! :)

> just to throw another idea out there... how might the UI change if the current label indicator ("annotate all occurrences of: car") itself became a control (i.e.dropdown list) .. would that be better restricted to the currently assigned labels (plus an entry for 'more..' which could take you to the add labels screen?), or what if it included everything in the dropdown list ..

cool idea! 

However, at the moment I am struggling a bit to find a good UI workflow. Imagine that we have such a dropdown in place. How can the user advance to the next annotation task? The easiest solution would probably be to keep that image and get the next annotation task from the queue. But I imagine that it could be a bit frustrating to see that image x times in a row with a different label (especially if it's an image that's not your prefered type of image; e.q stills vs crowded images). Another possibility would be to let the user switch between the list of labels. But would a switch already imply a "done"? And what about the "done" button - doesn't it get obsolete in that case?

**Comment by dobkeratops** [comment]
>But I imagine that it could be a bit frustrating to see that image x times in a row with a different la bel

fair enough.. it's possible the dropdown idea would conflict with the task-queue .. I wonder if this can be resolved

In 'labelme' it's more a case of looking for parts of the image that are un-labelled - given they have completely free labelling, so it's reasonable to expect they can cover the entire image. then again that's less precise if you're using bounding boxes.

What might be more ideal is rather than a concept of a task being completed, a progressive advancement of labelling resolution .. e.g. you might start out with rough bounding boxes, but you could always go in and improve that. Once you've labelled the 'dog'.. you go in and label the head, eyes, foreleg, rear-leg, tail ,paw... 

I wonder if it's possible to include a measure of error/fuzziness in the annotation (see also the idea of fuzzy blob annotations) .

regarding a task queue, with a sufficiently large pool of images and labels would it be safe to simply present images totally at random ? (again if you've got a rich set of parts - you can always improve on whats there... every window of every building, whatever..)

**Comment by dobkeratops** [comment]
On a positive note, one thing I definitely like about the current setup is the ability to specify labels without actually annotating - That is definitely in the spirit of "progressive refinement".  (in LabelMe you can freely assign, but you must actually click out a region to be able to say something). 

At the point of adding the label, you could say each pixel has a 1/N chance of being it (N =number of labels..). Then when you actually annotate any label, you narrow things down (probability of a pixel being an unlabelled label = 1/(N-(number of annotated labels)) )

Now imagine how that would behave with user selection of label - by picking the larger or easier labels first, people could narrow things across the image faster

**Comment by bbernhard** [comment]
> fair enough.. I can see the dropdown idea might conflict with the task-queue .. I wonder if this can be resolved.

The longer I think about it, the more I am in favor of a "pro mode". I think we could still keep the task based approach, if we build a different UI around it (if we combine labeling, validation and annotation in one view, one might not even notice). 

> What might be more ideal is rather than a concept of a task being completed, a progressive advancement of labelling resolution .. e.g. you might start out with rough bounding boxes, but you could always go in and improve that. Once you've labelled the 'dog'.. you go in and label the head, eyes, foreleg, rear-leg, tail ,paw...

definitely! 

> regarding a task queue, with a sufficiently large pool of images and labels would it be safe to simply present images totally at random ? (again if you've got a rich set of parts - you can always improve on whats there... every window of every building, whatever..)

> how about randomly picking images , preferring ones which currently have no annotations?

I have to check, because there are already a few places where we prefer images with no xxx. 
(e.q images that have no labels have preference over images with labels - so they are shown first, when in the label mode; but I am not sure if that also holds true for the annotation mode). 

> At the point of adding the label, you could say each pixel has a 1/N chance of being it (N =number of labels..). Then when you actually annotate any label, you narrow things down (probability of a pixel being an unlabelled label = 1/(N-(number of annotated labels)) )

> Now imagine how that would behave with user selection of label - by picking the easier labels first, people could narrow things down across the image faster; the difficult annotations are things like trees with fuzzy boundaries, intertwined objects, etc. (in those cases I like to be able to switch to annotating components, e.g. do the heads/hands first, or tree-trunks which have solid outlines.. etc)

Interesting idea! 

I am wondering if we could also use a neural net (Mask RCNN?) to do the ordering? e.q: we run the image through a neural net and check if any bounding rects/masks are found for the given labels. The biggest bounding rect/mask is probably the most dominant object in the picture. (that of course only works, if the model was trained on that label - so it might not work not for all labels; but over time it would be possible to retrain the model on new labels).

**Comment by dobkeratops** [comment]
r.e. mode jump:  would it be a comparable amount of work to add a label dropdown to the 'annotate' page? - I think that would get you more utility.  the other tweak I would suggest (which might be no UI work..) is to allow clicking on images in the explorer view to switch to annotate.. that would be really nice, scrolling through to find do-able tasks (after asking for a specific label)

**Comment by bbernhard** [comment]
> r.e. mode jump: would it be a comparable amount of work to add a label dropdown to the 'annotate' page? 

I think that shouldn't be that much work. But even if we have a label dropdown menu in place, how do we switch between the annotation tasks? (re. would a switch already imply a "done"? And what about the "done" button - doesn't it get obsolete in that case?)

What we _could_ probably do is the following: Each time you switch between a label a new canvas for drawing gets created. That means you could switch between the canvases back and forth to work on more annotation tasks at once and after you press the "done" button those changes are sent to the backend. Would that be a viable solution or does it get unnecessary complicated?

(To be honest I don't know how much work that would be...would need to look into that :). If it's too much work, it might be better to spend the time working on a pro mode? )

> allow clicking on images in the explorer view to switch to annotate.. that would be really nice, scrolling through to find do-able tasks (after asking for a specific label)

Definitely!

I am currently working on the "jump to validation"/"jump to labeling" switches, which should be done either tomorrow or the day after. 

Here's a short preview:

![jump_mode_validation](https://user-images.githubusercontent.com/5710631/41205726-1d04fec6-6cf9-11e8-911c-2bc358259303.gif)

Before I start with the "jump to annotation" I would like to push it to production in order to get some feedback (not sure if the current UIX workflow is the best one). If that works, I'll start working on the "jump to annotation" feature. 









btw: the next update also allows to change the annotation workspace size. 

![workspace_size](https://user-images.githubusercontent.com/5710631/41205830-dc25b222-6cfa-11e8-8b0d-da79ffa41157.gif)

**Comment by dobkeratops** [comment]
perhaps 'done' can be repurposed to 'next image'; also show any existing annotations for the selected label; - presumably you can go back over the same image and add annotations for a label if someone missed something, changing label would show the existing annotations ?

**Comment by dobkeratops** [comment]
> I am currently working on the "jump to validation"/"jump to labeling" switches, which should be done either by tomorrow or the day after.

interesting.. maybe even the mode jump could acheive this, i.e. what if you had some choice on which label it will use when jumping from 'add labels' to 'annotate' (the last one you added, or perhaps whatever is in the selection box at the time?)
